File: game/packages/engine/src/combat/turn-states.ts
/**
 * Turn phase states ported from TurnStates.cs.
 * Each state implements IState and operates on CombatState from shared types.
 */

import type { CombatState } from "@deckbuilder/shared";
import type { IState } from "./state-machine";

// ---------------------------------------------------------------------------
// Callback signatures used by the turn states.
// The combat manager wires these up so states stay decoupled from systems.
// ---------------------------------------------------------------------------

export interface TurnCallbacks {
  drawCards(state: CombatState, count: number): CombatState;
  enablePlayerInput(): void;
  disablePlayerInput(): void;
  resolveEnemyIntents(state: CombatState): CombatState;
  resolveEndOfTurn(state: CombatState): CombatState;
}

// ---------------------------------------------------------------------------
// Shared context that every turn state reads / writes.
// ---------------------------------------------------------------------------

export interface TurnContext {
  state: CombatState;
  callbacks: TurnCallbacks;
  /** Called by a state when it is finished and the FSM should advance. */
  onPhaseComplete: () => void;
}

// ---------------------------------------------------------------------------
// Draw Phase — draw N cards, then immediately signal completion.
// ---------------------------------------------------------------------------

export class DrawState implements IState {
  constructor(
    private readonly ctx: TurnContext,
    private readonly drawCount: number,
  ) {}

  enter(): void {
    this.ctx.state = this.ctx.callbacks.drawCards(this.ctx.state, this.drawCount);
    this.ctx.state = { ...this.ctx.state, turnPhase: "draw" };
  }

  update(): void {
    this.ctx.onPhaseComplete();
  }

  exit(): void {}
}

// ---------------------------------------------------------------------------
// Player Action — waits for the player to play cards / end turn.
// ---------------------------------------------------------------------------

export class PlayerActionState implements IState {
  constructor(private readonly ctx: TurnContext) {}

  enter(): void {
    this.ctx.state = { ...this.ctx.state, turnPhase: "player_action" };
    this.ctx.callbacks.enablePlayerInput();
  }

  update(): void {
    /* Waits for external command (card play / end turn button). */
  }

  exit(): void {
    this.ctx.callbacks.disablePlayerInput();
  }
}

// ---------------------------------------------------------------------------
// Enemy Intent — resolve all enemy actions, then signal completion.
// ---------------------------------------------------------------------------

export class EnemyIntentState implements IState {
  constructor(private readonly ctx: TurnContext) {}

  enter(): void {
    this.ctx.state = { ...this.ctx.state, turnPhase: "enemy_intent" };
    this.ctx.state = this.ctx.callbacks.resolveEnemyIntents(this.ctx.state);
  }

  update(): void {
    this.ctx.onPhaseComplete();
  }

  exit(): void {}
}

// ---------------------------------------------------------------------------
// Resolution — tick status effects, clean up, then signal completion.
// ---------------------------------------------------------------------------

export class ResolutionState implements IState {
  constructor(private readonly ctx: TurnContext) {}

  enter(): void {
    this.ctx.state = { ...this.ctx.state, turnPhase: "resolution" };
    this.ctx.state = this.ctx.callbacks.resolveEndOfTurn(this.ctx.state);
  }

  update(): void {
    this.ctx.onPhaseComplete();
  }

  exit(): void {}
}

// ---------------------------------------------------------------------------
// End Turn — bump the turn counter and signal completion so the loop restarts.
// ---------------------------------------------------------------------------

export class EndTurnState implements IState {
  constructor(private readonly ctx: TurnContext) {}

  enter(): void {
    this.ctx.state = {
      ...this.ctx.state,
      turnPhase: "end_turn",
      turnNumber: this.ctx.state.turnNumber + 1,
    };
  }

  update(): void {
    this.ctx.onPhaseComplete();
  }

  exit(): void {}
}

File: muse-weave/unity/DeckbuilderRoguelite/Assets/Scripts/Core/FSM/TurnStates.cs
using System;

namespace Deckbuilder.Core.FSM
{
    public class DrawPhaseState : IState
    {
        private readonly Action<int> _drawCards;
        private readonly int _drawCount;
        private readonly Action _onComplete;

        public DrawPhaseState(Action<int> drawCards, int drawCount, Action onComplete)
        {
            _drawCards = drawCards;
            _drawCount = drawCount;
            _onComplete = onComplete;
        }

        public void Enter() => _drawCards(_drawCount);
        public void Execute() => _onComplete();
        public void Exit() { }
    }

    public class PlayerActionState : IState
    {
        private readonly Action _enableInput;
        private readonly Action _disableInput;

        public PlayerActionState(Action enableInput, Action disableInput)
        {
            _enableInput = enableInput;
            _disableInput = disableInput;
        }

        public void Enter() => _enableInput();
        public void Execute() { /* Waits for player commands */ }
        public void Exit() => _disableInput();
    }

    public class EnemyIntentState : IState
    {
        private readonly Action _resolveEnemyIntents;
        private readonly Action _onComplete;

        public EnemyIntentState(Action resolveEnemyIntents, Action onComplete)
        {
            _resolveEnemyIntents = resolveEnemyIntents;
            _onComplete = onComplete;
        }

        public void Enter() => _resolveEnemyIntents();
        public void Execute() => _onComplete();
        public void Exit() { }
    }

    public class ResolutionState : IState
    {
        private readonly Action _resolveEndOfTurn;
        private readonly Action _onComplete;

        public ResolutionState(Action resolveEndOfTurn, Action onComplete)
        {
            _resolveEndOfTurn = resolveEndOfTurn;
            _onComplete = onComplete;
        }

        public void Enter() => _resolveEndOfTurn();
        public void Execute() => _onComplete();
        public void Exit() { }
    }

    public class TargetSelectionState : IState
    {
        private readonly Action _showTargetUI;
        private readonly Action _hideTargetUI;

        public TargetSelectionState(Action showTargetUI, Action hideTargetUI)
        {
            _showTargetUI = showTargetUI;
            _hideTargetUI = hideTargetUI;
        }

        public void Enter() => _showTargetUI();
        public void Execute() { /* Waits for target selection */ }
        public void Exit() => _hideTargetUI();
    }
}
